### PR TITLE
Added checkboxHorizontalMargin to DataTable and PaginatedDataTable

### LINKED ISSUE #71156: Allow us to change margin of the checkbox in a DataTable
## Use case

Currently the margin between the checkbox and content in the first column is the horizontalMargin:
https://api.flutter.dev/flutter/material/DataTable/horizontalMargin.html

I want to be able to change the margin for the checkbox independently.

## Proposal

Add new property to the DataTable and DataTableThemeData called `checkboxHorizontalMargin` to allow us to set this value independently of horizontalMargin.

### PR DESCRIPTION
**Pull request by chinmoy12c**
## Description

Added checkboxHorizontalMargin to DataTable and PaginatedDataTable

## Related Issues

Fixes #71156 

## Tests

I added the following tests:

- checkboxHorizontalMargin properly applied
- PaginatedDataTable custom horizontal padding - checkbox

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/data_table.dart
Comment (perclasson) [inline packages/flutter/lib/src/material/data_table.dart:693]: Make sure there's a space here between the comment and the backslash.

File: packages/flutter/lib/src/material/data_table.dart
Comment (perclasson) [inline packages/flutter/lib/src/material/data_table.dart:641]: Can you make this change so that the `checkboxHorizontalMargin` is independent of the `horizontalMargin`.

as I understand it here, we will have

```
checkboxHorizontalMargin | Checkbox | checkboxHorizontalMargin / 2 | horizontalMargin / 2 | firstCellContent | horionztalMargin
```

The usage I was thinking of was to make sure that we have it like this:

```
checkboxHorizontalMargin | Checkbox | checkboxHorizontalMargin | horizontalMargin  | firstCellContent | horionztalMargin
```

This means you must update the following lines too:
https://github.com/flutter/flutter/blob/66fabe01c8a12a88e221a4166b337f6ec8848868/packages/flutter/lib/src/material/data_table.dart#L966

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/paginated_data_table_test.dart:426]: I would prefer if you add a new test, and leave this test as it is, here that uses the `_customCheckboxHorizontalMargin` so that we can confirm that the old behavior and the new behavior works.

File: packages/flutter/test/material/data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/data_table_test.dart:581]: I would prefer if you add a new test, and leave this test as it is, here that uses the `_customCheckboxHorizontalMargin` so that we can confirm that the old behavior and the new behavior works.

**Review by perclasson — COMMENTED** [review]
Thanks for the PR, looks good but I would suggest adding a **new** test that tests that the margin is correct for the first checkbox column and the checkboxes themselves when the `checkboxHorizontalMargin` is set and leave most of the old tests as they are.

You also need to update the margin here:
https://github.com/flutter/flutter/blob/66fabe01c8a12a88e221a4166b337f6ec8848868/packages/flutter/lib/src/material/data_table.dart#L937

**Comment by chinmoy12c** [comment]
Thanks for the review @perclasson working on the changes. :)

File: packages/flutter/test/material/data_table_test.dart
Comment (chinmoy12c) [reply packages/flutter/test/material/data_table_test.dart]: @perclasson made the necessary changes :+1:

File: packages/flutter/test/material/data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/data_table_test.dart:1567]: nit: `// Custom checkbox padding.`

File: packages/flutter/test/material/data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/data_table_test.dart:1579]: nit: `// First column padding.`

File: packages/flutter/test/material/data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/data_table_test.dart:1581]: super nit, period at the end: `// DataTable wraps its DataCells in an Align widget.`

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/paginated_data_table_test.dart:866]: nit: Custom checkbox padding.

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (perclasson) [inline packages/flutter/test/material/paginated_data_table_test.dart:878]: nit: Custom first column padding.

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (perclasson) [inline packages/flutter/lib/src/material/paginated_data_table.dart:228]: I think this can be updated to (no need for template):
```
Horizontal margin around the checkbox, if it is displayed.

If null, then [horizontalMargin] is used as the margin between the edge
of the table and the checkbox, as well as the margin between the checkbox
and the content in the first data column.
```

I think you can update `horizontalMargin` too to make it clear how it is used together with the new property:
```
The horizontal margin between the edges of the table and the content
in the first and last cells of each row.

This value defaults to 24.0 to adhere to the Material Design specifications.

If [checkboxHorizontalMargin] is null, then [horizontalMargin] is also the 
margin between the edge of the table and the checkbox, as well as the
margin between the checkbox and the content in the first data column.
```

**Review by perclasson — COMMENTED** [review]
This looks great!! LGTM after the comments I've made are resolved, I'll also leave it to @HansMuller to review this before it can be merged.

File: packages/flutter/lib/src/material/data_table.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/data_table.dart:769]: Why does this default to `effectiveHorizontalMargin / 2`? The API docs imply that it will default to 24, not 12 right?

**Review by HansMuller — COMMENTED** [review]
This looks good, just one question about the dimensions.

File: packages/flutter/lib/src/material/data_table.dart
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:769]: It is not obvious, but the end margin of the checkbox is made up of effectiveCheckboxHorizontalMarginEnd (i.e. effectiveHorizontalMargin / 2.0 ) here and the other part here:
https://github.com/flutter/flutter/blob/e030ddfee8a62116c91c35c2339deee41410f243/packages/flutter/lib/src/material/data_table.dart#L1037

It ends up being the full effectiveHorizontalMargin.

Comment (chinmoy12c) [reply packages/flutter/lib/src/material/data_table.dart:769]: > It is not obvious, but the end margin of the checkbox is made up of effectiveCheckboxHorizontalMarginEnd (i.e. effectiveHorizontalMargin / 2.0 ) here and the other part here:
> https://github.com/flutter/flutter/blob/e030ddfee8a62116c91c35c2339deee41410f243/packages/flutter/lib/src/material/data_table.dart#L1037
> 
> It ends up being the full effectiveHorizontalMargin.

Yes that is indeed the case :smile:

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by HansMuller** [comment]
The framework_tests-misc-linux is shown as unfinished (started 4+ days ago) however it appears to have actually finished OK. I will land this now.

**Pull request merged**
